### ISSUE TITLE
"proto package names should not be mixed in the same directory" false positive

### ISSUE THREAD

**Issue by srenatus** [lyft/protoc-gen-star#24]
### 💥 Sup?

On [this branch](https://github.com/lyft/protoc-gen-star/compare/master...srenatus:sr/bug/pkg-name-clash), running the example protoc plugin in `testdata/protos` yields what I believe to be **a false positive package name clash**:

```console
$ env DEBUG=1 protoc -I . --example_out="plugins:../generated" bug/api/config/config.proto
registering plugin: jsonify
registering module:  printer
reading input
parsing input proto
parsing command-line params
evaluating multi-package mode
single package run
registering plugin: gatherer
initializing plugins
preparing official generator
generating official PGG PBs and gathering PG* AST
[jsonify] Initializing
[gatherer] Initializing
[gatherer][package:config] importPath: "github.com/lyft/protoc-gen-star/testdata/protos/bug/pkg/config"
[gatherer][file:bug/pkg/config/main.proto] f.GetPackage() = lyft.pgs.bar.config
[gatherer][file:bug/pkg/config/main.proto] pkg.ProtoName() = lyft.pgs.bar.config
[gatherer][package:config] importPath: "github.com/lyft/protoc-gen-star/testdata/protos/bug/pkg/config"
[gatherer][file:bug/pkg/config/secondary.proto] f.GetPackage() = lyft.pgs.bar.config
[gatherer][file:bug/pkg/config/secondary.proto] pkg.ProtoName() = lyft.pgs.bar.config
[jsonify] [bug/api/config/config.proto] [Fizz] implementing json.Marshaler/Unmarshaler interface
[gatherer][package:config1] importPath: "github.com/lyft/protoc-gen-star/testdata/protos/bug/api/config"
[gatherer][file:bug/api/config/config.proto] f.GetPackage() = lyft.pgs.api.config
[gatherer][file:bug/api/config/config.proto] pkg.ProtoName() = lyft.pgs.bar.config
2018/06/27 16:31:09 protoc-gen-go: error:[gatherer][file:bug/api/config/config.proto]proto package names should not be mixed in the same directory (lyft.pgs.bar.config, lyft.pgs.api.config)
--example_out: protoc-gen-example: Plugin failed with status code 1.
```

I don't believe the situation set up there is invalid. 😃 

ℹ️ This could be related to the changes done for protobuf v1.1.0, I haven't seen it before. Also, I believe parts of the relevant code paths have been touched recently.

**Comment by glerchundi** [comment]
Same here, I solved this by pinning my generator to pre 1.1.0 commit, concretely to this one: `6fc21c85aeae4a7f0ddfa6c79269bcc28972c8d1`.

**Comment by glerchundi** [comment]
@srenatus i started working on this this morning because i needed due to emerging conflicts with another vendored library requiring protobuf-1.1.0.

I'm pretty sure that the source of the problem is [this line](https://github.com/lyft/protoc-gen-star/blob/f3b83fca61817e55058744eb1a3c8e2e95a0dada/gatherer.go#L60) in combination with concrete package names. In case you have two packages with different proto package name origins but the same go package names one is going to overlap the other and therefore you will be hitting this issue.

For example.:

`a.proto`:
```
package a;

import "google/api/annotations.proto";
import "lyft/api/annotations.proto"
```

`google/api/annotations.proto`:
```
package google.api;
option go_package = "google.golang.org/genproto/googleapis/api/annotations;annotations";
[...]
```

`lyft/api/annotations.proto`:
```
package lyft.api;
option go_package = "lyft.golang.org/apis/api/annotations;annotations";
[...]
```

Mainly because the same go package name (different file source) will be stored in a [`pkgs`](https://github.com/lyft/protoc-gen-star/blob/f3b83fca61817e55058744eb1a3c8e2e95a0dada/gatherer.go#L94) map using the same key thus overriding an already existing `*pkg` instance.

Can you confirm this is happening to you too?

**Comment by srenatus** [comment]
I'll try verifying. Will get back to you in a moment.

**Comment by srenatus** [comment]
I think you're right. I've added printfs for `name` and `p`, and I find:
```
name = config
pkg = &{0xc4200b8b00 github.com/chef/a2/components/automate-config/pkg/config config [] }
name = config
name = config
2018/07/09 06:52:33 protoc-gen-go: error:[gatherer][file:components/automate-gateway/api/config/config_request.proto]proto package names should not be mixed in the same directory (chef.automate.infra.config, chef.automate.api.config)
--policy_out: protoc-gen-policy: Plugin failed with status code 1.
```

The first `name=config` is a proto file which has `go_package = "github.com/chef/a2/components/automate-config/pkg/config"`, the second one has `go_package = "github.com/chef/a2/components/automate-gateway/api/config"`.

**Comment by glerchundi** [comment]
Confirmed then, thanks! 👍 

I forgot to say that pinning to this sha fixes the issue: a7e273ba634fd50a2e358b334c62d0a09b7b657b. This is the first 1.1.0 compatible version of this project.

I don't have time to further investigate/propose solution, but now at least it's pretty clear which is the source of the problem.

**Comment by srenatus** [comment]
@glerchundi oh, right, pinning is a good idea. 0.3.4 is the version tag for that commit, so I'll go with that. thanks for looking into this.

**Comment by srenatus** [comment]
I think I was mistaken with the version tag there...

**Comment by glerchundi** [comment]
I think the solution I proposed in #25 is the easiest way to solve this one.

PTAL @rodaine @btc

**Comment by glerchundi** [comment]
I'm being bitten constantly with this issue :(

@srenatus in case you're using versioning in your go package names like this: `"mycompany.com/apis/dummy-go/v1;dummy";` the sha i told you was working...it's not. Package name resolves to `v1` instead of `dummy`. It seems like the commits after that one fixed the issue with this concrete package resolving, so with the PR i recently sent, all this nightmare should be solved, I hope.

In case you need this fix:
```
[[constraint]]
  name = "github.com/lyft/protoc-gen-star"
  source = "https://github.com/glerchundi/protoc-gen-star.git"
  revision = "db9f39e97b811423e7591b9f7ac1e03a6a518880"
```

**Comment by rodaine** [comment]
Fixed in #30
